### PR TITLE
plugin/https: Add max_streams to limit HTTP/2 concurrent streams

### LINKED ISSUE #8521: plugin/https: Make HTTP/2 max concurrent streams configurable
**What would you like to be added**:

Allow operators to configure the maximum number of concurrent HTTP/2
streams per connection on the *https* (DoH) server, via a `max_streams`
option in the `https {}` block.

**Why is this needed**:

The DoH server currently relies on the Go HTTP/2 server's built-in
default for the maximum number of concurrent streams per connection, with
no way to tune it from the Corefile.

In high-performance and high-fan-in deployments this is limiting in both
directions:

- **Clients that multiplex heavily over a single connection.** When many
  downstream clients share one persistent DoH connection (for example a
  relay/forwarder that pools requests, or an edge gateway), request
  concurrency on that single connection can approach the stream limit
  during bursts. Being able to raise the limit lets such a connection
  absorb the burst instead of queuing streams.

- **Resource protection.** Conversely, operators who want a tighter bound
  per connection (to cap memory/goroutine usage from a single client)
  currently cannot lower it.

Today the only lever exposed is `max_connections`; there is no equivalent
for per-connection stream concurrency. The sibling *https3* plugin already
exposes `max_streams` for QUIC, so DoH lacks parity.

**Proposed behavior**

Add `max_streams` to the `https {}` block, mirroring the *https3*
plugin's semantics:

- omitted  → use the Go HTTP/2 server default
- `0`      → use the underlying HTTP/2 transport default
- positive → advertise exactly that many concurrent streams
- negative → rejected at config parse time

Example:

```
https://.:443 {
   tls cert.pem key.pem
   https {
       max_connections 0
       max_streams 512
   }
   forward . 10.0.0.10
}
```

**Additional context**

This keeps the *https* and *https3* plugins consistent and gives
operators a single, transport-appropriate knob for tuning per-connection
concurrency for their workload.

### PR DESCRIPTION
**Pull request by myohannes**
### 1. Why is this pull request needed and what does it do?

The *https* (DoH) server relies on the Go HTTP/2 server's built-in default
for the maximum number of concurrent streams per connection, with no way to
tune it from the Corefile. This adds a `max_streams` option to the `https {}`
block so operators can:

- **Raise** the per-connection stream limit for high-fan-in clients that
  multiplex many requests over a single persistent DoH connection (e.g. a
  relay/forwarder or edge gateway), so a burst is absorbed instead of queued.
- **Lower** it to cap memory/goroutine usage from a single client.

Behavior mirrors the existing *https3* plugin's `max_streams`:
- omitted  → Go HTTP/2 server default is used
- `0`      → use the underlying HTTP/2 transport default
- positive → advertise exactly that many concurrent streams
- negative → rejected at config parse time

The limit is applied via the standard library `http.Server.HTTP2`
(`HTTP2Config.MaxConcurrentStreams`), so it is advertised in the server's
HTTP/2 SETTINGS frame.

### 2. Which issues (if any) are related?

Fixes #8521

### 3. Which documentation changes (if any) need to be made?

Updated `plugin/https/README.md` to document the new `max_streams` option and
its semantics (including `0` = transport default).

### 4. Does this introduce a backward incompatible change or deprecation?

No. `max_streams` is a new, optional directive. When it is not specified the
behavior is unchanged (the Go HTTP/2 server default is used).

### CODE REVIEW

File: core/dnsserver/config.go [hunk 1]
```diff
@@ -114,6 +114,10 @@ type Config struct {
 	// This is nil if not specified, allowing for a default to be used.
 	MaxHTTPSConnections *int
 
+	// MaxHTTPSStreams defines the maximum number of concurrent HTTP/2 streams per HTTPS connection.
+	// This is nil if not specified, allowing for a default to be used.
+	MaxHTTPSStreams *int
```
Comment (houyuwushang) [inline core/dnsserver/config.go:119]: [P2] Propagate the new setting to every key in a server block

Please copy `MaxHTTPSStreams` in `propagateConfigParams`. Caddy executes directive setup only for the first key, so this valid multi-transport block stores the setting on the DNS config but leaves the HTTPS config nil:

```corefile
.:0 https://.:0 {
    bind 127.0.0.1
    tls ../plugin/tls/test_cert.pem ../plugin/tls/test_key.pem
    https {
        max_streams 7
    }
    whoami
}
```

Using the existing test certificates, I measured `SETTINGS_MAX_CONCURRENT_STREAMS=250`; reversing the two keys produces 7. The configured limit must not depend on key order. A Corefile-based regression would catch this, whereas the direct `NewServerHTTPS` tests bypass propagation.

File: core/dnsserver/server_https.go [hunk 2]
```diff
@@ -91,6 +95,19 @@ func NewServerHTTPS(addr string, group []*Config) (*ServerHTTPS, error) {
 		IdleTimeout:  s.IdleTimeout,
 		ErrorLog:     stdlog.New(&loggerAdapter{}, "", 0),
 	}
+	// max_streams limits the number of concurrent HTTP/2 streams per connection. When unset,
+	// DefaultHTTPSMaxStreams is applied; a value of 0 leaves the underlying HTTP/2 transport
+	// default in place; a positive value sets the limit explicitly. The chosen value is
+	// advertised in the server's SETTINGS frame.
+	maxStreams := DefaultHTTPSMaxStreams
+	if len(group) > 0 && group[0] != nil && group[0].MaxHTTPSStreams != nil {
+		maxStreams = *group[0].MaxHTTPSStreams
```
Comment (houyuwushang) [inline core/dnsserver/server_https.go]: [P2] Resolve explicit limits across the shared listener group

Reading only `group[0]` also silently ignores a setting in a later server block. I started two blocks on the same listener: `https://first.example.:0` without an `https` directive, followed by `https://second.example.:0` with `max_streams 7`. The connection advertises 250; reversing the blocks makes it advertise 7. Both configurations start successfully.

Please resolve the explicit setting across the group and reject incompatible values, or reject unsupported placements rather than silently discarding them. A shared HTTP/2 connection needs one well-defined listener-level limit. Please cover both block orders, omitted versus explicit values, and conflicting values.

File: plugin/https/setup.go [hunk 3]
```diff
@@ -54,6 +54,22 @@ func parseDOH(c *caddy.Controller) error {
...
+			if len(args) != 1 {
+				return c.ArgErr()
+			}
+			val, err := strconv.Atoi(args[0])
+			if err != nil {
+				return c.Errf("invalid max_streams value '%s': %v", args[0], err)
+			}
+			if val < 0 {
+				return c.Errf("max_streams must be a non-negative integer: %d", val)
+			}
```
Comment (houyuwushang) [inline plugin/https/setup.go:69]: [P3, non-blocking] Reject values outside the HTTP/2 setting's range

On a 64-bit build, `strconv.Atoi` accepts `max_streams 4294967297`, but Go converts `HTTP2Config.MaxConcurrentStreams` to `uint32` internally. I measured an advertised limit of 1 for that configuration, rather than a startup error. Please reject values above `math.MaxUint32` before assigning the option, and add an upper-bound parser test. This is a low-priority invalid-input edge case, separate from the configuration-loss issues above.

**Review by houyuwushang — CHANGES_REQUESTED** [review]
I checked this through real TLS/HTTP/2 connections. The single-listener path advertises the configured value and rejects excess concurrent streams, but two supported Corefile arrangements silently fall back to 250. Please address the configuration propagation and shared-listener cases below and cover them through Corefile-to-wire tests. The existing focused tests pass; the new probes reproduce both cases consistently. The integer-range comment is lower priority.

File: core/dnsserver/config.go (hunk 1 above)
Comment (myohannes) [reply core/dnsserver/config.go:119]: Done. propagateConfigParams now copies MaxHTTPSStreams, so the limit no longer depends on key order. Added a Corefile-based regression in test/https_test.go (TestHTTPSMaxStreamsKeyOrder) using your exact .:0 https://.:0 block, both key orders, asserting the advertised SETTINGS_MAX_CONCURRENT_STREAMS on the wire — it reproduces 250 before the fix and 7 after, for either order. Also added TestPropagateConfigParamsMaxHTTPSStreams as a direct unit test.

File: plugin/https/setup.go (hunk 3 above)
Comment (myohannes) [reply plugin/https/setup.go:69]: Fixed. max_streams > math.MaxUint32 is now rejected at parse time before assignment. Added upper-bound parser tests in TestSetup: 4294967295 accepted, 4294967296 rejected with a "must not exceed" error.

File: core/dnsserver/server_https.go (hunk 2 above)
Comment (myohannes) [reply core/dnsserver/server_https.go]: Fixed to resolve max_streams across the whole group instead of reading group[0], and to reject conflicting explicit values at startup with a clear error. Covered by TestNewServerHTTPSStreamsAcrossGroup: value on the first member, on the second member, omitted on all members, and conflicting values.

File: plugin/https/setup.go [hunk 4]
```diff
@@ -54,6 +55,25 @@ func parseDOH(c *caddy.Controller) error {
...
+			if val > math.MaxUint32 {
```
Comment (houyuwushang) [inline plugin/https/setup.go:70]: [P1] Keep the upper-bound check and its tests portable to 32-bit targets

The [release job](https://github.com/coredns/coredns/actions/runs/33938427962/job/101252512895) fails on Linux/ARM because `val` is an `int`: this comparison requires the constant 4294967295 to fit in a 32-bit int. I reproduced it with `GOOS=linux GOARCH=arm CGO_ENABLED=0 go build ./plugin/https`. Widening the comparison after the negative-value check fixes the production build:

```suggestion
			if int64(val) > math.MaxUint32 {
```

Please also make the two new boundary cases in `TestSetup` architecture-aware. After changing this guard, ARM test compilation still fails at `intPtr(4294967295)` in `setup_test.go:124`. Also, on 32-bit builds `Atoi("4294967296")` returns an `invalid max_streams` error before reaching this guard, rather than `must not exceed`. Use `strconv.IntSize` to distinguish the expected acceptance/error cases and avoid an out-of-range int constant. I verified the comparison plus adjusted fixtures with Linux/ARM test compilation and Windows/386 test execution.

File: test/https_test.go [hunk 5]
```diff
@@ -3,12 +3,14 @@ package test
...
 	"crypto/tls"
+	"encoding/binary"
 	"io"
 	"net"
 	"net/http"
 	"testing"
 	"time"
 
+	"github.com/coredns/caddy"
 	"github.com/miekg/dns"
```
Comment (houyuwushang) [inline test/https_test.go:14]: [P2] Separate CoreDNS imports from third-party imports

The [e2e job](https://github.com/coredns/coredns/actions/runs/33938427962/job/101252512877) fails in `TestImportOrdering` because `github.com/coredns/caddy` and `github.com/miekg/dns` are in the same import group. The repository's checker requires a blank line between them:

```suggestion
	"github.com/coredns/caddy"

	"github.com/miekg/dns"
```

I reproduced this with the repository's import-order checker and verified that this separator clears the error for `https_test.go`.

**Review by houyuwushang — CHANGES_REQUESTED** [review]
The behavior fixes address my earlier comments. Both key orders and shared-listener orders now advertise the configured limit, explicit zero is handled consistently, conflicting values are rejected, and oversized input is rejected on 64-bit builds. I re-ran these wire cases with race detection. Two CI blockers remain below; I validated the suggested fixes locally, including ARM test compilation and an actual 32-bit parser test run.

File: plugin/https/setup.go (hunk 4 above)
Comment (myohannes) [reply plugin/https/setup.go]: Thanks for the detailed review. Widened the guard to int64(val) > math.MaxUint32 so the Linux/ARM release build compiles. Made the two TestSetup boundary cases architecture-aware: the value is derived at runtime via strconv.Atoi instead of an intPtr(4294967295) literal (which overflowed a 32-bit int at compile time). On 32-bit, Atoi rejects both 4294967295 and 4294967296 as invalid max_streams before the guard, so the expectations branch on that.

Verified on real ARM hardware (aarch64): ARM prod build, native 32-bit GOARCH=arm TestSetup execution, and TestImportOrdering all pass.

File: test/https_test.go (hunk 5 above)
Comment (myohannes) [reply test/https_test.go:15]: Fixed. Added a blank line between github.com/coredns/caddy and github.com/miekg/dns in test/https_test.go.

**Review by houyuwushang — COMMENTED** [review]
Both remaining findings are fixed in c335603f9, and I have resolved the threads. I independently checked the ARM build and test compilation, executed the 32-bit parser tests, and rechecked the Corefile-to-HTTP/2 behavior on the current-master merge. No new code issues found. The latest GitHub Actions workflows still need approval to run; I will leave final approval until the latest-head CI, including e2e and release, completes.

**Review by houyuwushang — APPROVED** [review]
The latest-head e2e and multi-architecture release/Docker checks now pass as well. The portability and import-order fixes address the remaining review findings, with no further issues from my re-review. LGTM.
